File: generator/dart/generate_json.dart
import 'dart:convert';
import 'dart:io';

import 'const.dart';
import 'utils.dart';

Map<String, Map<String, Object?>> parameter_infos = {};
Map<String, Map<String, Object?>> abc_classes = {};
Map<String, Map<String, Object?>> classes = {};
Map<String, Map<String, Object?>> functions = {};

void reset() {
  parameter_infos.clear();
}


(String, Map<String, Object?>, String) process_schema(String raw_body) {
  raw_body = raw_body.replaceAll(';', '').trim();
  var data = raw_body.split("=");
  var class_body = data[0].trim().split(' ');
  var parent = data[1];

  var class_name = class_body[0].toCamelCase;

  class_body.sublist(1).map((e) {
    processTlParameters(e, parameter_infos);
  }).toList();
  return (class_name.trim(), parameter_infos, parent.trim());
}

String? process_docs(String raw_comment) {
  String _comment = "";
  for (var r in RegExp(r"(?=(@(param_)?(\w+)(.+?))(@|$))", multiLine: true)
      .allMatches(raw_comment)) {
    var parameter = r.group(3)!;
    var description = r.group(4);
    var parameter_ = r.group(2);

    if (parameter == 'description' && parameter_ == null)
      // print(description)
      _comment = description?.trim() ?? _comment;
    else {
      parameter_infos[parameter] = {
        "description": description?.trim(),
        "nullable":
            RegExp(r"(may be null|If empty|pass null)", caseSensitive: false)
                .hasMatch(description ?? '')
      };
    }
  }
  return _comment;
}

void put_abc(String key, {String? value, String? description}) {
  var data = abc_classes[key.trim()];
  // print(data);
  assert(value != null || description != null);
  if (data == null) {
    data = {"description": description, "child": Set()};
    // print(data);
    if (value != null) ((data["child"]) as Set).add(value.trim());
    abc_classes[key.trim()] = data;
    return;
  }
  (data["child"] as Set).add(value!.trim());
}

void generate_json() {
  var data = File(TL_API_FILE).readAsStringSync();
  bool isFunction = false;
  var regex = RegExp(
    r"(//@class (.+) @description (.+))|(//@description(.|\n)+?)(\n.* = .*;)$",
    multiLine: true,
  );
  var _class_or_func = classes;
  var iterable = regex.allMatches(data);
  for (var i in iterable) {
    final abc = i.group(2);
    final abc_desc = i.group(3);
    final docs = i.group(4);
    final schema = i.group(6);

    if (abc != null && abc_desc != null) {
      put_abc(abc, description: abc_desc);
    } else {
      if (schema ==
          "\ngetAuthorizationState = AuthorizationState;") // now function starts
      {
        isFunction = true;
        _class_or_func = functions;
      }
      var description = process_docs(docs!);
      var (class_name, parameter_infos, parent) =
          process_schema(schema!);
      var _data = {
        "description": description,
        "parameters": json.decode(json.encode(parameter_infos)),
        "return": null,
        "parent": null
      };

      if (isFunction) {
        _data["return"] = parent;
      }
      else {
        if (parent.toLowerCase() != class_name.toLowerCase())
          _data['parent'] = parent;
        else
          _data['parent'] = 'TlObject';
      }

      _class_or_func[class_name] = _data; //add class or function to dictionary
      if (!isFunction && parent.toLowerCase() != class_name.toLowerCase())
        put_abc(parent, value: class_name);
      reset();
    }
  }
}

void main() {
  generate_json();
  var test = json.decode(File(ABC_CLASS_JSON_FILE).readAsStringSync());
  abc_classes.forEach((key, value) {
    test[key]!;
    assert((test[key]['child'] as List).length == (value['child'] as Set).length);
  });
  // print(json.encode(
  //   abc_classes,
  //   toEncodable: (object) {
  //     if (object is Set) {
  //       return object.toList();
  //     }
  //   },
  // ));

  test = json.decode(File(CLASS_JSON_FILE).readAsStringSync());
  classes.forEach((key, value) {
    test[key]!;
    assert((test[key]['parameters'] as Map).length == (value['parameters'] as Map).length);
  });
  //   print(json.encode(
  //   classes,
  //   toEncodable: (object) {
  //     if (object is Set) {
  //       return object.toList();
  //     }
  //   },
  // ));
  test = json.decode(File(FUNC_JSON_FILE).readAsStringSync());
  functions.forEach((key, value) {
    test[key]!;
    assert((test[key]['parameters'] as Map).length == (value['parameters'] as Map).length);
  });
  // print(json.encode(
  //   functions,
  //   toEncodable: (object) {
  //     if (object is Set) {
  //       return object.toList();
  //     }
  //   },
  // ));


  
}

File: generator/generate_json.py
import json
import re
import os
from typing import Optional

from const import *
from utils import CamelCase, Serializer, get_dart_type

parameter_descriptions = {}
abc_classes = {}
classes = {}
functions = {}


def reset():
    # print(parameter_descriptions)
    parameter_descriptions.clear()
    # input()
    ...

def get_tl_to_dart(tl: str):
    parameter, _type = tl.split(":")
    _type, vectorElementType, _enum, depth = get_dart_type(_type)
    info = {
        "type": _type,
        "vectorElementType": vectorElementType,
        "enum": _enum,
        "vector_depth": depth
    }
    return parameter, info

def process_tl_parameter(source: str, lookup_dict: dict):
    parameter, data = get_tl_to_dart(source)
    result: Optional[dict] = lookup_dict.get(parameter, None)
    if not result:
        raise KeyError(f"parameter: {parameter} is not available in lookup_dict")
    result: dict
    result.update(data)
    return {parameter: result}



def process_comment(raw_comment: str):
    _comment = None
    for r in re.finditer(r"(?=(@(param_)?(\w+)(.+?))(@|$))", raw_comment.strip(), re.M):
        parameter, description = r.group(3), r.group(4)  # variable # description
        parameter_ = r.group(2)  # param_
        if parameter == 'description' and parameter_ is None:
            # print(description)
            _comment = description
        else:
            parameter_descriptions[parameter] = {
                "description": description.strip(),
                "nullable": bool(re.search("(?i)(may be null|If empty|pass null)", description))
            }
    return _comment.strip()


def process_schema(schema: str):
    class_body, parent = schema.strip()[:-1].split("=")  # ignore ';' and split
    class_body = class_body.split()
    # print(class_body)

    class_name = CamelCase(class_body[0])
    class_param = map(lambda a: process_tl_parameter(a, parameter_descriptions), class_body[1:])
    list(class_param)
    return class_name, parameter_descriptions, parent.strip()


def put_abc(key, value=None, description=None):
    data = abc_classes.get(key.strip(), None)
    assert value or description
    if not data:
        data = {
            "description": description,
            "child": set()
        }
        if value:
            data["child"].add(value.strip())
        abc_classes[key.strip()] = data
        return
    data["child"].add(value.strip())


def generate_json():
    with open(TL_API_FILE) as f:
        is_function = False
        regex = re.compile(r"(//@class (.+) @description (.+))|(//@description(.|\n)+?)(\n.* = .*;)$", re.M)
        iterable = regex.finditer(f.read())
        _class_or_func = classes
        for i in iterable:
            # assert len(i.groups()) == 3, (i.groups(), len(i.groups()))
            _, abc, abc_desc, raw_comment, _, schema = i.groups()
        #   1,   2,     3,          4,     5,    6, 

            # assert not abc and not abc_desc
            if abc and abc_desc:  # abstract class
                # print(abc, abc_desc, sep=" : ")
                put_abc(abc, description=abc_desc)
                # continue
            else:

                if schema == "\ngetAuthorizationState = AuthorizationState;":  # now function starts
                    is_function = True
                    _class_or_func = functions
                description = process_comment(raw_comment)
                class_name, class_body, parent = process_schema(schema)

                _data = {
                    "description": description,
                    "parameters": class_body.copy(),
                    "return": None,
                    "parent": None,
                }
                if is_function:
                    issync = bool(re.search('Can be called synchronously', description, re.I))
                    _data['parent'] = 'SyncFunc' if issync else 'Func'
                    _data["return"] = parent
                else:
                    if parent.lower() != class_name.lower():
                        _data['parent'] = parent
                    else:
                        _data['parent'] = 'TlObject'

                # print(_data)
                _class_or_func[class_name] = _data  # add class or function to dictionary

                if not is_function and parent.lower() != class_name.lower():
                    put_abc(parent, class_name)

                # print()
                # print(comment)
                # print(class_name, class_body, sep=" : ")
                # print()
                reset()
    os.makedirs(BASE_DIR_JSON, exist_ok=True)

    for name, data in zip(
            (CLASS_JSON_FILE, ABC_CLASS_JSON_FILE, FUNC_JSON_FILE),
            (classes, abc_classes, functions)
    ):
        with open(name, "w") as f:
            json.dump(data, f, cls=Serializer, indent=4)

    return abc_classes, classes, functions


if __name__ == "__main__":
    generate_json()
